### DISCUSSION TITLE
printing double values and using mathematical functions do not work on Cooja

### DISCUSSION THREAD

**Discussion by LouisDubly** [contiki-ng/contiki-ng#2922] [Q&A]
Hello,
I try to print double values on cooja but i have this output:
![image](https://github.com/contiki-ng/contiki-ng/assets/119162693/e1473050-8aa7-467a-97ab-17df95329569)
Here is my code:
#include "contiki.h"
#include "net/routing/routing.h"
#include "math.h"
#include "net/netstack.h"
#include "net/ipv6/simple-udp.h"
#include "stdint.h"
#include "inttypes.h"
#include "sys/log.h"
#include <stdio.h>

#define LOG_MODULE "App"
#define LOG_LEVEL LOG_LEVEL_INFO

#define UDP_CLIENT_PORT    8765
#define UDP_SERVER_PORT    5678

#define L 100 

static struct simple_udp_connection udp_conn;
static uint32_t tx_count = 0;
static uint32_t total_bits_sent = 0;  
double tau = 0.3; 
double lambda;
double U;

struct packet {
  uint32_t seq;
  char data[(L / 8) + sizeof(uint32_t)];
};

#define PACKET_QUEUE_SIZE 100 
static struct packet packet_queue[PACKET_QUEUE_SIZE];
static uint8_t queue_head = 0;
static uint8_t queue_tail = 0;  

double my_exp(double x) {
    double result = 1.0;
    double term = 1.0;
    int i;
    
    for (i = 1; i < 10; ++i) { 
        term *= x / i;
        result += term;
    }
    
    return result;
}

static void enqueue_packet(uint32_t seq, const char* data) {
  memcpy(packet_queue[queue_tail].data + sizeof(uint32_t), data, L / 8);
  memcpy(packet_queue[queue_tail].data, &seq, sizeof(uint32_t));
  packet_queue[queue_tail].seq = seq;
  queue_tail = (queue_tail + 1) % PACKET_QUEUE_SIZE;
}

static void send_next_packet(uip_ipaddr_t* dest_ipaddr) {
  if (queue_head != queue_tail) {
    const struct packet* pkt = &packet_queue[queue_head];
    LOG_INFO("Sending request %"PRIu32" to ", pkt->seq);
    LOG_INFO_6ADDR(dest_ipaddr);
    LOG_INFO_("\n");
    simple_udp_sendto(&udp_conn, pkt->data, sizeof(pkt->data), dest_ipaddr);
    queue_head = (queue_head + 1) % PACKET_QUEUE_SIZE;
    total_bits_sent += L;
    LOG_INFO("%lu,%"PRIu32"\n", clock_seconds(), pkt->seq);
  }
}

/*---------------------------------------------------------------------------*/
PROCESS(udp_client_process, "UDP client");
AUTOSTART_PROCESSES(&udp_client_process);
/*---------------------------------------------------------------------------*/
static void
udp_rx_callback(struct simple_udp_connection *c,
         const uip_ipaddr_t *sender_addr,
         uint16_t sender_port,
         const uip_ipaddr_t *receiver_addr,
         uint16_t receiver_port,
         const uint8_t *data,
         uint16_t datalen)
{
}

/*---------------------------------------------------------------------------*/
PROCESS_THREAD(udp_client_process, ev, data)
{
  static struct etimer periodic_timer;
  static struct etimer stats_timer;
  static char str[L / 8] = "hello";
  uip_ipaddr_t dest_ipaddr;
  static uint32_t missed_tx_count;

  PROCESS_BEGIN();

  LOG_INFO("Starting UDP client\n");

  LOG_INFO("Initial tau: %f\n", tau);

  /* Initialize UDP connection */
  simple_udp_register(&udp_conn, UDP_CLIENT_PORT, NULL,
                      UDP_SERVER_PORT, udp_rx_callback);

  etimer_set(&periodic_timer, CLOCK_SECOND);  
  etimer_set(&stats_timer, CLOCK_SECOND);   

  while(1) {
    PROCESS_WAIT_EVENT();

    if (etimer_expired(&periodic_timer)) {
      if (NETSTACK_ROUTING.node_is_reachable() &&
          NETSTACK_ROUTING.get_root_ipaddr(&dest_ipaddr)) {
        enqueue_packet(tx_count, str);
        send_next_packet(&dest_ipaddr);
        tx_count++;
      } else {
        LOG_INFO("Not reachable yet\n");
        if (tx_count > 0) {
          missed_tx_count++;
        }
      }
      etimer_reset(&periodic_timer);
    }

    if (etimer_expired(&stats_timer)) {
      double Beta = my_exp(tau); 

      lambda = 1 / Beta;
      U = (lambda * L) / 250000.0;  

      LOG_INFO("tau: %f, Beta: %f, lambda: %f, U: %f\n", tau, Beta, lambda, U);

      etimer_reset(&stats_timer);
    }
  }

  PROCESS_END();
}

Moreover, instead of creating an exponential function, i firstly tried to use the mathematical function exp() but i got this error everytime while compiling on cooja:
undefined reference to ´exp´

Can someone please help me on this?

**Comment by alexrayne** [comment]
looks that you uses old contiki?
from version 4.7 they switch contiki log-printer of cooja platform on a standart libc printer, exactly for printing floats.

on older vefsions of contiki floats not supported. you can try this [PR](https://github.com/arurke/contiki-ng/pull/4)  proposed float printers for embedded targets, and cooja too. there, for use float printers, maybe you need configure like `#define PRINTF_CONF_HAVE_DOUBLE 1` on your project config

**Reply by LouisDubly** [reply]
Hello, thank you for your reply,
I tried to PR what you sent me, but it didn't work.
I use contiki-ng 4.9
Do you have another idea of how to solve this problem?

**Reply by alexrayne** [reply]
"I tried to PR what you sent me, but it didn't work." - that is clarify diagnostics ))

**Comment by rahav** [comment]
See the solution [here](https://github.com/contiki-ng/contiki-ng/issues/2566#issue-1827151804)

**Reply by LouisDubly** [reply]
Hello,
I saw that I had to switch back to Contiki 4.7, but I didn't understand how to do it. Could you please explain?
Thanks in advance.

**Reply by rahav** [reply]
You can use Contiki 4.9 if you apply the [patch](https://github.com/contiki-ng/contiki-ng/pull/2567) by Mr. Jonsson.

Prior to the patch I developed a crud library as a workaround. Let me know if above does not work for you and I will see if I can find and post it.

**Reply by alexrayne** [reply]
> See the solution [here](https://github.com/contiki-ng/contiki-ng/issues/2566#issue-1827151804)

That was merged last year into release

**Reply by LouisDubly** [reply]
Unfortunately, it looks like the commit is not existing anymore
![image](https://github.com/contiki-ng/contiki-ng/assets/119162693/0a48fa14-9b21-499d-ac60-991949dace85)
So i have made the changes manually on the files, but it didnt worked for me.

**Reply by rahav** [reply]
Are you interested in the workaround I mentioned?

**Reply by LouisDubly** [reply]
Yes please, if its ok for you

**Reply by rahav** [reply]
````c
/* Return either the Whole part of a real number, or its Decimal part.

   This is a rudimentary function, required as a workaround Contiki limitation. Contiki knows how to print %f only
   in Cooja or when the TARGET=native. When compiled to run on hardware, it does not. See
   https://github.com/contiki-ng/contiki-ng/discussions/2703

   The function is generally used in LOG_ or printf() statements. For example, suppose that we want to print a
   probability pr, with 2 decimal digits accuracy. Such a number can be between 0.00 to 1.00. We can print it using
   the following:

		LOG_INFO("Probability = %d.%0*u\n",
			get_number_part(pr, 0),
			2, get_number_part(pr, 2));

   Note that in the format specifier we used %0* to make it show leading 0 and have variable length. The * will be
   replaced by the value 2, which is provided as an argument to the format specification.

   Known limitations:
	- *** WARNING *** : Will wrongly return a positive whole part, when the whole part equals 0.
	- Limited to returning up to 3 decimal digits. If higher precision is needed, add another `else if` in the
		precision calculation.
        - DO NOT CHANGE the types of the real parameter, or the fraction var. This may lead to loss of accuracy. 

   Params:
	real A real number from which the whole or decimal part is to be extracted.
	precision 	0 Return the whole part.
			1 Return 1 digit of the decimal part.
			2 Return 2 digits.
			3 Return 3 digits (the default).

   @returns the Whole or Decimal part of real.
 */
static int get_number_part(double real, unsigned char precision) {
	double fraction;

	if (precision == 0) {
		return (int)real;		
	}

	fraction = real - (int)real;
	if (fraction < 0) {
		fraction *= -1.0;
	}

	if (precision == 1) {
		return (int)(fraction * 10.0);
	} else if (precision == 2) {
		return (int)(fraction * 100.0);
	}
	return (int)(fraction * 1000.0);
}
````
